### PR TITLE
TST: restore type checks to maybe_promote tests

### PR DESCRIPTION
**Pull request by jbrockmendel**
xref #25425.  Working on fixing the xfailed tests in this file.  This is going to be a marathon.

### CODE REVIEW

**Comment by h-vetinari** [comment]
Very cool - I didn't expect that the typecheck could be re-introduced so easily. Thanks for tackling this!

OTOH, most of worst tests of #23982 were turned into dummies for #25637. [As I](https://github.com/pandas-dev/pandas/pull/25637#issuecomment-504488421) [offered](https://github.com/pandas-dev/pandas/pull/25425#pullrequestreview-255888873), I can rebase #23982, so we have the same baseline of tests available (somewhere *outside* #25424).

File: pandas/core/dtypes/cast.py [hunk 1]
```diff
@@ -404,6 +411,25 @@ def maybe_promote(dtype, fill_value=np.nan):
     return dtype, fill_value
 
 
+def _check_lossless_cast(value, dtype: np.dtype) -> bool:
+    """
+    Check if we can cast the given value to the given dtype _losslesly_.
```
Comment (gfyoung) [inline pandas/core/dtypes/cast.py]:
```suggestion
    Check if we can cast the given value to the given dtype _losslessly_.
```

File: pandas/core/dtypes/cast.py [hunk 2]
```diff
@@ -404,6 +411,25 @@ def maybe_promote(dtype, fill_value=np.nan):
...
+    dtype : np.dtype
+
+    Returns
+    -------
+    bool
+    """
+    casted = dtype.type(value)
+    if casted == value:
+        return True
+    return False
```
Comment (gfyoung) [inline pandas/core/dtypes/cast.py]: Why not just `return dtype.type(value) == value` ?

Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py]: Either way would work.  I think in future steps this function is going to have to check a couple more things.

File: pandas/core/dtypes/cast.py [hunk 3]
```diff
@@ -366,6 +367,8 @@ def maybe_promote(dtype, fill_value=np.nan):
     elif is_bool(fill_value):
         if not issubclass(dtype.type, np.bool_):
             dtype = np.object_
+        else:
+            fill_value = np.bool_(fill_value)
```
Comment (jreback) [inline pandas/core/dtypes/cast.py]: you added some code here but are these actually hit?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/cast.py (hunk 3 above)
Comment (jreback) [reply pandas/core/dtypes/cast.py]: i see you have another PR on top; can you just do them in one

Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py]: > you added some code here but are these actually hit?

Yes.  All the changes to the non-test code are needed to make the changes in the test code pass.

Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py]: > i see you have another PR on top; can you just do them in one

Sure.  Just close this and move conversation there?

Comment (jreback) [reply pandas/core/dtypes/cast.py]: yeah let's just do it in one

**Comment by jbrockmendel** [comment]
Closing in favor of #28564.

**Pull request closed without merging**